In [15]:
!pip install pattern
!pip install nltk
!pip install scikit-learn
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

# Task1 (Extracting Western Subsection from URL)

In [16]:
from pattern.web import URL, plaintext

url0='https://en.wikipedia.org/wiki/Emoticon'
print(url0)
byt = URL(url0).download() # byte object
s = str(byt,'utf-8')       # byte to string
s = plaintext(s, keep={'h1':[], 'h2':[], 'h3':[]})
print(s)

https://en.wikipedia.org/wiki/Emoticon
Emoticon - Wikipedia

<h1>Emoticon</h1>

From Wikipedia, the free encyclopedia

This is the latest accepted revision, reviewed on 13 February 2022.

Jump to navigationJump to search

Pictorial representation of a facial expression using punctuation marks, numbers and letters

Not to be confused with Emoji or Sticker (messaging).

For technical reasons, search terms like ";;" redirect here.

This article contains Unicode emoticons or emojis. Without proper rendering support, you may see question marks, boxes, or other symbols instead of the intended characters.

A smiley-face emoticon

Examples of kaomoji smileys

An emoticon (/ɪˈmoʊtɪkɒn/, ə-MOH-tə-kon, rarely pronounced /ɪˈmɒtɪkɒn/, ih-MAW-tih-kon),[1][2][3][4] short for "emotion icon",[5] also known simply as an emote,[citation needed] is a pictorial representation of a facial expression using characters—usually punctuation marks, numbers, and letters—to express a person's feelings, mood or reac

In [17]:
import re

In [18]:
section_name = "Styles"
subsection_name = "Western"

In [19]:
header2_pattern = re.compile("<h2>\s*(.*)\s*\[edit\]\s*</h2>")

In [20]:
h2s = header2_pattern.finditer(s)

In [21]:
style_section_match = None
next_section_match = None
for h2 in h2s:
    if h2.group(1) == "Styles":
        style_section_match = h2
    elif style_section_match is not None and next_section_match is None:
        next_section_match = h2

In [22]:
style_section_text = ""
if next_section_match is None:
    style_section_text = s[style_section_match.end():]
else:
    style_section_text = s[style_section_match.end():next_section_match.start()]

In [23]:
len(style_section_text)

12622

In [24]:
header3_pattern = re.compile("<h3>\s*(.*)\s*\[edit\]\s*</h3>")

In [25]:
h3s = header3_pattern.finditer(style_section_text)

In [26]:
western_section_match = None
next_section_match = None
for h3 in h3s:
    if h3.group(1) == "Western":
        western_section_match = h3
    elif western_section_match is not None and next_section_match is None:
        next_section_match = h3

In [27]:
western_section_text = ""
if next_section_match is None:
    western_section_text = s[style_section_match.end()+western_section_match.end():]
else:
    western_section_text = s[style_section_match.end()+western_section_match.end():style_section_match.end()+next_section_match.start()]

In [28]:
western_section_text

'\n\nUsually, emoticons in Western style have the eyes on the left, followed by the nose and the mouth. The two-character version\n:)\n\nwhich omits the nose is also very popular.\n\nThe most basic emoticons are relatively consistent in form, but each of them can be transformed by being rotated (making them tiny ambigrams), with or without a hyphen (nose).\nThere are also some possible variations to emoticons to get new definitions, like changing a character to express a new feeling, or slightly change the mood of the emoticon. For example,\n:(\n\nequals sad and\n:((\n\nequals very sad. Weeping can be written as\n:\'(\n\n. A blush can be expressed as\n:">\n\n. Others include wink\n;)\n\n, a grin\n:D\n\n, smug\n:->\n\n, and can be used to denote a flirting or joking tone, or may be implying a second meaning in the sentence preceding it.[48]\n;P\n\n, such as when blowing a raspberry. An often used combination is also\n<3\n\nfor a heart, and\nfor a broken heart.\n:O\n\nis also sometimes u

In [29]:
print('Length of subsection "Western":', len(western_section_text))

Length of subsection "Western": 4026


In [30]:
re.compile("[.!?:;]").split(western_section_text)

['\n\nUsually, emoticons in Western style have the eyes on the left, followed by the nose and the mouth',
 ' The two-character version\n',
 ')\n\nwhich omits the nose is also very popular',
 '\n\nThe most basic emoticons are relatively consistent in form, but each of them can be transformed by being rotated (making them tiny ambigrams), with or without a hyphen (nose)',
 '\nThere are also some possible variations to emoticons to get new definitions, like changing a character to express a new feeling, or slightly change the mood of the emoticon',
 ' For example,\n',
 '(\n\nequals sad and\n',
 '((\n\nequals very sad',
 ' Weeping can be written as\n',
 "'(\n\n",
 ' A blush can be expressed as\n',
 '">\n\n',
 ' Others include wink\n',
 ')\n\n, a grin\n',
 'D\n\n, smug\n',
 '->\n\n, and can be used to denote a flirting or joking tone, or may be implying a second meaning in the sentence preceding it',
 '[48]\n',
 'P\n\n, such as when blowing a raspberry',
 ' An often used combination is also

In [34]:
import nltk
from nltk.tokenize import sent_tokenize
nltk.download("punkt")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [35]:
sentences = sent_tokenize(western_section_text)

In [36]:
whitespace_tokens = []
for sentence in sentences:
    whitespace_tokens.append(re.compile("\s").split(sentence))

In [37]:
whitespace_tokens

[['',
  '',
  'Usually,',
  'emoticons',
  'in',
  'Western',
  'style',
  'have',
  'the',
  'eyes',
  'on',
  'the',
  'left,',
  'followed',
  'by',
  'the',
  'nose',
  'and',
  'the',
  'mouth.'],
 ['The',
  'two-character',
  'version',
  ':)',
  '',
  'which',
  'omits',
  'the',
  'nose',
  'is',
  'also',
  'very',
  'popular.'],
 ['The',
  'most',
  'basic',
  'emoticons',
  'are',
  'relatively',
  'consistent',
  'in',
  'form,',
  'but',
  'each',
  'of',
  'them',
  'can',
  'be',
  'transformed',
  'by',
  'being',
  'rotated',
  '(making',
  'them',
  'tiny',
  'ambigrams),',
  'with',
  'or',
  'without',
  'a',
  'hyphen',
  '(nose).'],
 ['There',
  'are',
  'also',
  'some',
  'possible',
  'variations',
  'to',
  'emoticons',
  'to',
  'get',
  'new',
  'definitions,',
  'like',
  'changing',
  'a',
  'character',
  'to',
  'express',
  'a',
  'new',
  'feeling,',
  'or',
  'slightly',
  'change',
  'the',
  'mood',
  'of',
  'the',
  'emoticon.'],
 ['For',
  'examp

# Task 2 (Different types of Tokenizers)

In [38]:
from nltk.tokenize import SpaceTokenizer
from nltk.tokenize import TweetTokenizer
from nltk.tokenize import word_tokenize
from nltk.tokenize import TreebankWordTokenizer

In [39]:
word_tokenizer_tokens = []
for sentence in sentences:
    word_tokenizer_tokens.append(word_tokenize(sentence))

In [40]:
word_tokenizer_tokens

[['Usually',
  ',',
  'emoticons',
  'in',
  'Western',
  'style',
  'have',
  'the',
  'eyes',
  'on',
  'the',
  'left',
  ',',
  'followed',
  'by',
  'the',
  'nose',
  'and',
  'the',
  'mouth',
  '.'],
 ['The',
  'two-character',
  'version',
  ':',
  ')',
  'which',
  'omits',
  'the',
  'nose',
  'is',
  'also',
  'very',
  'popular',
  '.'],
 ['The',
  'most',
  'basic',
  'emoticons',
  'are',
  'relatively',
  'consistent',
  'in',
  'form',
  ',',
  'but',
  'each',
  'of',
  'them',
  'can',
  'be',
  'transformed',
  'by',
  'being',
  'rotated',
  '(',
  'making',
  'them',
  'tiny',
  'ambigrams',
  ')',
  ',',
  'with',
  'or',
  'without',
  'a',
  'hyphen',
  '(',
  'nose',
  ')',
  '.'],
 ['There',
  'are',
  'also',
  'some',
  'possible',
  'variations',
  'to',
  'emoticons',
  'to',
  'get',
  'new',
  'definitions',
  ',',
  'like',
  'changing',
  'a',
  'character',
  'to',
  'express',
  'a',
  'new',
  'feeling',
  ',',
  'or',
  'slightly',
  'change',
  '

In [41]:
st = SpaceTokenizer()

In [42]:
st_tokenizer_tokens = []
for sentence in sentences:
    st_tokenizer_tokens.append(st.tokenize(sentence))

In [43]:
st_tokenizer_tokens

[['\n\nUsually,',
  'emoticons',
  'in',
  'Western',
  'style',
  'have',
  'the',
  'eyes',
  'on',
  'the',
  'left,',
  'followed',
  'by',
  'the',
  'nose',
  'and',
  'the',
  'mouth.'],
 ['The',
  'two-character',
  'version\n:)\n\nwhich',
  'omits',
  'the',
  'nose',
  'is',
  'also',
  'very',
  'popular.'],
 ['The',
  'most',
  'basic',
  'emoticons',
  'are',
  'relatively',
  'consistent',
  'in',
  'form,',
  'but',
  'each',
  'of',
  'them',
  'can',
  'be',
  'transformed',
  'by',
  'being',
  'rotated',
  '(making',
  'them',
  'tiny',
  'ambigrams),',
  'with',
  'or',
  'without',
  'a',
  'hyphen',
  '(nose).'],
 ['There',
  'are',
  'also',
  'some',
  'possible',
  'variations',
  'to',
  'emoticons',
  'to',
  'get',
  'new',
  'definitions,',
  'like',
  'changing',
  'a',
  'character',
  'to',
  'express',
  'a',
  'new',
  'feeling,',
  'or',
  'slightly',
  'change',
  'the',
  'mood',
  'of',
  'the',
  'emoticon.'],
 ['For',
  'example,\n:(\n\nequals',


In [50]:
tt = TweetTokenizer()

In [51]:
tt_tokenizer_tokens = []
for sentence in sentences:
    tt_tokenizer_tokens.append(tt.tokenize(sentence))

In [54]:
tt_tokenizer_tokens

[['Usually',
  ',',
  'emoticons',
  'in',
  'Western',
  'style',
  'have',
  'the',
  'eyes',
  'on',
  'the',
  'left',
  ',',
  'followed',
  'by',
  'the',
  'nose',
  'and',
  'the',
  'mouth',
  '.'],
 ['The',
  'two-character',
  'version',
  ':)',
  'which',
  'omits',
  'the',
  'nose',
  'is',
  'also',
  'very',
  'popular',
  '.'],
 ['The',
  'most',
  'basic',
  'emoticons',
  'are',
  'relatively',
  'consistent',
  'in',
  'form',
  ',',
  'but',
  'each',
  'of',
  'them',
  'can',
  'be',
  'transformed',
  'by',
  'being',
  'rotated',
  '(',
  'making',
  'them',
  'tiny',
  'ambigrams',
  ')',
  ',',
  'with',
  'or',
  'without',
  'a',
  'hyphen',
  '(',
  'nose',
  ')',
  '.'],
 ['There',
  'are',
  'also',
  'some',
  'possible',
  'variations',
  'to',
  'emoticons',
  'to',
  'get',
  'new',
  'definitions',
  ',',
  'like',
  'changing',
  'a',
  'character',
  'to',
  'express',
  'a',
  'new',
  'feeling',
  ',',
  'or',
  'slightly',
  'change',
  'the',


In [52]:
tree = TreebankWordTokenizer()

In [53]:
tree_tokenizer_tokens = []
for sentence in sentences:
    tree_tokenizer_tokens.append(tree.tokenize(sentence))

In [55]:
tree_tokenizer_tokens

[['Usually',
  ',',
  'emoticons',
  'in',
  'Western',
  'style',
  'have',
  'the',
  'eyes',
  'on',
  'the',
  'left',
  ',',
  'followed',
  'by',
  'the',
  'nose',
  'and',
  'the',
  'mouth',
  '.'],
 ['The',
  'two-character',
  'version',
  ':',
  ')',
  'which',
  'omits',
  'the',
  'nose',
  'is',
  'also',
  'very',
  'popular',
  '.'],
 ['The',
  'most',
  'basic',
  'emoticons',
  'are',
  'relatively',
  'consistent',
  'in',
  'form',
  ',',
  'but',
  'each',
  'of',
  'them',
  'can',
  'be',
  'transformed',
  'by',
  'being',
  'rotated',
  '(',
  'making',
  'them',
  'tiny',
  'ambigrams',
  ')',
  ',',
  'with',
  'or',
  'without',
  'a',
  'hyphen',
  '(',
  'nose',
  ')',
  '.'],
 ['There',
  'are',
  'also',
  'some',
  'possible',
  'variations',
  'to',
  'emoticons',
  'to',
  'get',
  'new',
  'definitions',
  ',',
  'like',
  'changing',
  'a',
  'character',
  'to',
  'express',
  'a',
  'new',
  'feeling',
  ',',
  'or',
  'slightly',
  'change',
  '

# Task 3 (Count Vectorizer)

In [44]:
from sklearn.feature_extraction.text import CountVectorizer

In [58]:
cvec = CountVectorizer(min_df=1, tokenizer=tt.tokenize)

# sents turned into sparse vector of word frequency counts
sents_counts = cvec.fit_transform(sentences)
# foovec now contains vocab dictionary which maps unique words to indexes
print('vocabulary mapped to numbers\n',cvec.vocabulary_)

vocabulary mapped to numbers
 {'usually': 292, ',': 4, 'emoticons': 108, 'in': 143, 'western': 301, 'style': 256, 'have': 138, 'the': 267, 'eyes': 117, 'on': 197, 'left': 166, 'followed': 122, 'by': 73, 'nose': 187, 'and': 48, 'mouth': 182, '.': 6, 'two-character': 281, 'version': 297, ':)': 15, 'which': 305, 'omits': 195, 'is': 153, 'also': 43, 'very': 298, 'popular': 211, 'most': 181, 'basic': 59, 'are': 54, 'relatively': 222, 'consistent': 87, 'form': 125, 'but': 72, 'each': 104, 'of': 191, 'them': 268, 'can': 74, 'be': 60, 'transformed': 279, 'being': 62, 'rotated': 229, '(': 2, 'making': 174, 'tiny': 274, 'ambigrams': 45, ')': 3, 'with': 307, 'or': 201, 'without': 308, 'a': 36, 'hyphen': 141, 'there': 269, 'some': 251, 'possible': 213, 'variations': 296, 'to': 275, 'get': 129, 'new': 185, 'definitions': 93, 'like': 170, 'changing': 78, 'character': 79, 'express': 114, 'feeling': 120, 'slightly': 248, 'change': 77, 'mood': 179, 'emoticon': 107, 'for': 124, 'example': 113, ':(': 14,

In [59]:
print('word counts for each sentence\n',sents_counts.toarray())

word counts for each sentence
 [[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 2 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 1 0]]


In [60]:
from sklearn.feature_extraction.text import TfidfTransformer
tfidf_transformer = TfidfTransformer()
sents_tfidf = tfidf_transformer.fit_transform(sents_counts)
print('word tfidf scores for each sentence\n',sents_tfidf.toarray())

word tfidf scores for each sentence
 [[0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.29983746 ... 0.         0.         0.        ]
 ...
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.31940806 0.        ]]


### Western 301 index

## Most and Least important words in Corpus

In [ ]:
vocab = cvec.get_feature_names()
vocab_idf_values = tfidf_transformer.idf_
v = list(zip(vocab,vocab_idf_values))
print(sorted(v,key=lambda l:l[1],reverse=True))